frontend/catalyst/python_interface/transforms/quantum/parity_synth.py

```diff
@@ -0,0 +1,532 @@
+
+
+def make_phase_polynomial(
+    ops: list[CustomOp],
+    init_wire_map: dict[QubitType, int],
+) -> tuple[np.ndarray]:
+    r"""Compute the phase polynomial representation of a list of ``CustomOp``\ s.
+    This implementation is very similar to :func:`~.transforms.intermediate_reps.phase_polynomial`
+    but adjusted to work with xDSL objects."""
+    wire_map = init_wire_map
+
+    parity_matrix = np.eye(len(wire_map), dtype=int)
+    parity_table = []
+    angles = []
+    arith_ops = []
+    for op in ops:
+        name = op.gate_name.data
+        if name == "CNOT":
+            control, target = wire_map.pop(op.in_qubits[0]), wire_map.pop(op.in_qubits[1])
+            parity_matrix[target] += parity_matrix[control]
+            wire_map[op.out_qubits[0]] = control
+            wire_map[op.out_qubits[1]] = target
+            continue
+
+        # RZ
+        angle = op.operands[0]
+        if getattr(op, "adjoint", False):
+            neg_op = arith.NegfOp(angle)
+            arith_ops.append(neg_op)
+            angle = neg_op.result
+        angles.append(angle)
+        wire = wire_map.pop(op.in_qubits[0])
+        parity_table.append(parity_matrix[wire].copy())  # append _current_ parity (hence the copy)
+        wire_map[op.out_qubits[0]] = wire
+
+    return parity_matrix % 2, np.array(parity_table).T % 2, angles, arith_ops
+
+
+def _cnot(i: int, j: int, inv_wire_map: dict[int, QubitType]):
+    """Create a CNOT operator acting on the qubits that map to wires ``i`` and ``j``
+    and update the wire map so that ``i`` and ``j`` point to the output qubits afterwards."""
+    cnot_op = CustomOp(
+        in_qubits=[inv_wire_map[i], inv_wire_map[j]],
+        gate_name="CNOT",
+        params=tuple(),
+    )
+    inv_wire_map[i] = cnot_op.out_qubits[0]
+    inv_wire_map[j] = cnot_op.out_qubits[1]
+    return cnot_op
+
+
+def _rz(wire: int, angle: SSAValue[builtin.Float64Type], inv_wire_map: dict[int, QubitType]):
+    """Create a CNOT operator acting on the qubit that maps to ``wire``
+    and update the wire map so that ``wire`` points to the output qubit afterwards."""
+    rz_op = CustomOp(in_qubits=[inv_wire_map[wire]], gate_name="RZ", params=(angle,))
+    inv_wire_map[wire] = rz_op.out_qubits[0]
+    return rz_op
+
+
+class ParitySynthPattern(pattern_rewriter.RewritePattern):
+    """Rewrite pattern that applies ``ParitySynth`` to subcircuits that constitute
+    phase polynomials.
+    """
+
+    phase_polynomial_ops: list[CustomOp]
+    init_wire_map: [QubitType, int]
+    global_wire_map: [QubitType, int]
+    phase_polynomial_ops: set[QubitType]
+    num_phase_polynomial_qubits: int
+
+    def __init__(self, *args, **kwargs):
+        if not has_networkx:  # pragma: no cover
+            raise ModuleNotFoundError(
+                "The packages networkx and galois are required to run the ParitySynth pass."
+                "You can install them via ``pip install networkx galois``."
```

Comment:
Why is `galois` mentioned here if it's not used? Is it being used under the hood by networkx?

Reply:
`galois` is used by the rowcol functionality. This is resolved in [this PR](https://github.com/PennyLaneAI/pennylane/pull/8771) though :)

Reply:
should update to remove galois from this error message